astropy/table/array.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+
+class ArraySlot(object):
+    __lt__ = lambda x, y: x.key < y.key
+    __le__ = lambda x, y: x.key <= y.key
+    __eq__ = lambda x, y: x.key == y.key
+    __ge__ = lambda x, y: x.key >= y.key
+    __gt__ = lambda x, y: x.key > y.key
+    __ne__ = lambda x, y: x.key != y.key
+
+    def __init__(self, key, val):
+        self.key = key
+        self.data = val
+    def __repr__(self):
+        return str((self.key, self.data))
+    def __str__(self):
+        return repr(self)
+
+class SortedArray(object):
+    '''
+    Implements a sorted array container using
+    a numpy structured array.
+    '''
+
+    def __init__(self, lines, col_dtypes=None, data=None):
```

Comment:
The names are a bit confusing or non-informative to me.  I think that `lines` should be more like `key_cols_table` (if I'm understanding).  I really don't get what `data` is, because `data is None` case `data` ends up being a list of columns + the index column, but in the block below you set `self.length = len(data)` => length of the list => n_key_columns + 1.